### PR TITLE
[lvgl] base implementation

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

Implement minimal LVGL base. Only supported widgets are `label` and `obj`.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3678

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [x] RP2040
- [x] host

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
esphome:
  name: lvgl

host:

logger:
  level: DEBUG
  logs:
    lvgl: DEBUG
    sensor: INFO
    touchscreen: INFO

display:
  - platform: sdl
    update_interval: never
    auto_clear_enabled: false
    dimensions:
      width: 450
      height: 600

lvgl:
  widgets:
    - label:
        text: Hello world
        text_color: 0xFF8000
        align: center
        text_font: montserrat_40
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/core/defines.h [hunk 1]
```diff
@@ -38,6 +38,10 @@
 #define USE_LIGHT
 #define USE_LOCK
 #define USE_LOGGER
+#define USE_LVGL
+#define LVGL_USES_FONT
+#define LVGL_USES_IMAGE
+#define LV_USE_TOUCHSCREEN
```
Comment (jesserockz) [inline esphome/core/defines.h]: These should follow the style of all others please =)
```suggestion
#define USE_LVGL_FONT
#define USE_LVGL_IMAGE
#define USE_LVGL_TOUCHSCREEN
```

File: esphome/components/lvgl/__init__.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                LOGGER.warning(
+                    "Using auto_clear_enabled: true in display config not recommended with LVGL"
+                )
+    buffer_frac = config[CONF_BUFFER_SIZE]
+    if not CORE.is_host and buffer_frac > 0.5 and "psram" not in global_config:
+        LOGGER.warning("buffer_size: may need to be reduced without PSRAM")
+
+
+async def to_code(config):
+    warning_checks(config)
```
Comment (jesserockz) [inline esphome/components/lvgl/__init__.py]: This should really be called in a `FINAL_VALIDATION_SCHEMA`

File: esphome/components/lvgl/__init__.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                filter(lambda c, k=display_id: c[CONF_ID] == k, display_conf)
+            )[0]
+            if CONF_LAMBDA in display:
+                LOGGER.warning(
+                    "Using lambda: in display config not recommended with LVGL"
+                )
+            if display[CONF_AUTO_CLEAR_ENABLED]:
+                LOGGER.warning(
+                    "Using auto_clear_enabled: true in display config not recommended with LVGL"
+                )
```
Comment (jesserockz) [inline esphome/components/lvgl/__init__.py]: Is there a case to allow this to be set to true? If not, I would raise an Invalid error. Same with the lambda above.

File: esphome/components/lvgl/widget.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+    @staticmethod
+    def set_completed():
+        Widget.widgets_completed = True
+
+    def __init__(self, var, wtype: WidgetType, config: dict = None, parent=None):
+        self.var = var
+        self.type = wtype
+        self.config = config
+        self.parent = None
```
Comment (jesserockz) [inline esphome/components/lvgl/widget.py]: This is immediately being set below
```suggestion
```

File: esphome/components/lvgl/__init__.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    for font in helpers.esphome_fonts_used:
+        await cg.get_variable(font)
+        cg.new_Pvariable(ID(f"{font}_engine", True, type=FontEngine), MockObj(font))
+
+    with LvContext():
+        await set_obj_properties(lv_scr_act, config)
+        if widgets := config.get(df.CONF_WIDGETS):
+            for w in widgets:
+                lv_w_type, w_cnfig = next(iter(w.items()))
+                await widget_to_code(w_cnfig, lv_w_type, lv_scr_act.obj)
```
Comment (jesserockz) [inline esphome/components/lvgl/__init__.py]:
```suggestion
        for w in config.get(df.CONF_WIDGETS, []):
            lv_w_type, w_cnfig = next(iter(w.items()))
            await widget_to_code(w_cnfig, lv_w_type, lv_scr_act.obj)
```

File: esphome/components/lvgl/widget.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+    """
+    Converts a Widget definition to C code.
+    :param w_cnfig: The widget configuration
+    :param w_type:  The Widget type
+    :param parent: The parent to which the widget should be added
+    :return:
+    """
+    spec: WidgetType = WIDGET_TYPES.get(w_type)
+    if not spec:
+        raise cv.Invalid(f"No handler for widget {w_type}")
```
Comment (jesserockz) [inline esphome/components/lvgl/widget.py]: Invalid cannot be raised inside `to_code` functions as it just results in a not so user friendly crash. If it is possible to get an invalid widget type, then more needs to be done at the validation stage.

File: esphome/components/lvgl/widget.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+    else:
+        var = cg.Pvariable(wid, cg.nullptr, type_=lv_obj_t)
+        lv_assign(var, creator)
+
+    widget = Widget.create(wid, var, spec, w_cnfig, parent)
+    await set_obj_properties(widget, w_cnfig)
+    if widgets := w_cnfig.get(CONF_WIDGETS):
+        for w in widgets:
+            sub_type, sub_config = next(iter(w.items()))
+            await widget_to_code(sub_config, sub_type, widget.obj)
```
Comment (jesserockz) [inline esphome/components/lvgl/widget.py]:
```suggestion
    for w in w_cnfig.get(CONF_WIDGETS, []):
        sub_type, sub_config = next(iter(w.items()))
        await widget_to_code(sub_config, sub_type, widget.obj)
```

File: esphome/components/lvgl/widget.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+    """
+    Add all widgets to an object
+    :param parent: The enclosing obj
+    :param config: The configuration
+    :return:
+    """
+    if widgets := config.get(CONF_WIDGETS):
+        for w in widgets:
+            w_type, w_cnfig = next(iter(w.items()))
+            await widget_to_code(w_cnfig, w_type, parent.obj)
```
Comment (jesserockz) [inline esphome/components/lvgl/widget.py]:
```suggestion
    for w in config.get(CONF_WIDGETS, []):
        w_type, w_cnfig = next(iter(w.items()))
        await widget_to_code(w_cnfig, w_type, parent.obj)
```

File: esphome/components/lvgl/widget.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+    widget = Widget.create(wid, var, spec, w_cnfig, parent)
+    await set_obj_properties(widget, w_cnfig)
+    if widgets := w_cnfig.get(CONF_WIDGETS):
+        for w in widgets:
+            sub_type, sub_config = next(iter(w.items()))
+            await widget_to_code(sub_config, sub_type, widget.obj)
+    await spec.to_code(widget, w_cnfig)
+
+
+async def add_widgets(parent: Widget, config: dict):
```
Comment (jesserockz) [inline esphome/components/lvgl/widget.py]: You have this function, but then are just duplicating the code? ~6 lines above here and in the init file...

File: esphome/components/lvgl/lvgl_esphome.h [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+#include "esphome/components/display/display.h"
+#include "esphome/components/display/display_color_utils.h"
+#include "esphome/core/component.h"
+#include "esphome/core/hal.h"
+#include "esphome/core/log.h"
+#include <lvgl.h>
+#include <vector>
+
+#ifdef LVGL_USES_FONT
```
Comment (jesserockz) [inline esphome/components/lvgl/lvgl_esphome.h]:
```suggestion
#ifdef USES_LVGL_FONT
```

File: esphome/components/lvgl/lvgl_esphome.h [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+#include "esphome/components/font/font.h"
+#endif
+namespace esphome {
+namespace lvgl {
+
+extern lv_event_code_t lv_custom_event;  // NOLINT
+#ifdef LVGL_USES_COLOR
```
Comment (jesserockz) [inline esphome/components/lvgl/lvgl_esphome.h]:
```suggestion
#ifdef USES_LVGL_COLOR
```

File: esphome/components/lvgl/lvgl_esphome.h [hunk 12]
```diff
@@ -0,0 +1,132 @@
+#pragma once
+#include "esphome/core/defines.h"
+#ifdef USE_LVGL
```
Comment (jesserockz) [inline esphome/components/lvgl/lvgl_esphome.h:3]: Probably don't need this as clang will run correctly with the library available, and any projects not using it wont copy the code to the generated project.
```suggestion
```

File: esphome/components/lvgl/lvgl_esphome.cpp [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+#include "lvgl_esphome.h"
+
+static const char *const TAG = "lvgl";
+
+namespace esphome {
+namespace lvgl {
+
+lv_event_code_t lv_custom_event;  // NOLINT
+void LvglComponent::setup() {
+  esph_log_config(TAG, "LVGL Setup starts");
```
Comment (jesserockz) [inline esphome/components/lvgl/lvgl_esphome.cpp]: Please stick to the defined macros so there is consistency throughout the codebase.
```suggestion
  ESP_LOGCONFIG(TAG, "LVGL Setup starts");
```

File: esphome/components/lvgl/lvgl_esphome.cpp [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+#include "esphome/core/log.h"
+#include "esphome/core/helpers.h"
+#include "esphome/core/hal.h"
+#include "lvgl_hal.h"
+#include "lvgl_esphome.h"
+
+static const char *const TAG = "lvgl";
+
+namespace esphome {
+namespace lvgl {
```
Comment (jesserockz) [inline esphome/components/lvgl/lvgl_esphome.cpp]: This should be inside the namespace really.
The functions outside can access it using `esphome::lvgl::TAG` correct?
```suggestion
namespace esphome {
namespace lvgl {

static const char *const TAG = "lvgl";
```

File: esphome/components/lvgl/lvgl_esphome.cpp [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+      break;
+  }
+  display->set_rotation(display::DISPLAY_ROTATION_0_DEGREES);
+  this->disp_drv_.hor_res = (lv_coord_t) display->get_width();
+  this->disp_drv_.ver_res = (lv_coord_t) display->get_height();
+  esph_log_v(TAG, "sw_rotate = %d, rotated=%d", this->disp_drv_.sw_rotate, this->disp_drv_.rotated);
+  this->disp_ = lv_disp_drv_register(&this->disp_drv_);
+  for (const auto &v : this->init_lambdas_)
+    v(this->disp_);
+  // this->display_->set_writer([](display::Display &d) { lv_timer_handler(); });
```
Comment (jesserockz) [inline esphome/components/lvgl/lvgl_esphome.cpp]: No commented out code please
```suggestion
```

**Review by jesserockz — DISMISSED** [review]

File: esphome/components/lvgl/lvgl_esphome.h (hunk 12 above)
Comment (clydebarrow) [reply esphome/components/lvgl/lvgl_esphome.h:3]: Pretty sure CI fails without this. Will suck it and see

File: esphome/components/lvgl/__init__.py [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+        result.append(display_list)
+    return result
+
+
+def final_validation(config):
+    global_config = full_config.get()
+    for display_id in get_display_list(config):
+        display = list(
+            filter(lambda c, k=display_id: c[CONF_ID] == k, global_config[CONF_DISPLAY])
+        )[0]
```
Comment (jesserockz) [inline esphome/components/lvgl/__init__.py]:
```suggestion
        path = global_config.get_path_for_id(display_id)[:-1]
        display = global_config.get_config_for_path(path)
```

File: esphome/components/lvgl/__init__.py [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+    definitions.sort()
+    return LV_CONF_H_FORMAT.format("\n".join(definitions))
+
+
+def get_display_list(config):
+    result = []
+    display_list = config.get(df.CONF_DISPLAYS)
+    if isinstance(display_list, list):
+        for display in config.get(df.CONF_DISPLAYS, []):
+            result.append(display[CONF_DISPLAY_ID])
```
Comment (jesserockz) [inline esphome/components/lvgl/__init__.py]:
```suggestion
        result.extend(display_list)
```

File: esphome/components/lvgl/__init__.py [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+            cv.GenerateID(df.CONF_DISPLAYS): cv.Any(
+                cv.use_id(Display),
+                cv.ensure_list(
+                    cv.maybe_simple_value(
+                        {
+                            cv.Required(CONF_DISPLAY_ID): cv.use_id(Display),
+                        },
+                        key=CONF_DISPLAY_ID,
+                    ),
+                ),
```
Comment (jesserockz) [inline esphome/components/lvgl/__init__.py]: Does multi display really need a `display_id` key for each one?
A simple list of ids can work too (goes with my above suggestion in `get_display_list`)
```suggestion
                cv.use_id(Display), cv.ensure_list(cv.use_id(Display))
```

File: esphome/components/lvgl/__init__.py [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+FINAL_VALIDATE_SCHEMA = final_validation
+
+CONFIG_SCHEMA = (
+    cv.polling_component_schema("1s")
+    .extend(obj_schema("obj"))
+    .extend(
+        {
+            cv.Optional(CONF_ID, default=df.CONF_LVGL_COMPONENT): cv.declare_id(
+                LvglComponent
+            ),
```
Comment (jesserockz) [inline esphome/components/lvgl/__init__.py]: Lets keep this simple and use the generated id like other components.
```suggestion
            cv.GenerateID(CONF_ID): cv.declare_id(LvglComponent),
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**